### ISSUE TITLE
Deep Dive Into OCR for Receipt Recognition - DZone AI

### ISSUE THREAD

**Issue by guevara** [guevara/read-it-later#8626]
Deep Dive Into OCR for Receipt Recognition - DZone AI<br>
<br>
https://ift.tt/xoBaJRw<br>
<br>
Jun. 21, 17<br>
<br>
<div>
<div>
<div>
<div>
<div>
<p dir="ltr">Optical Character Recognition is a process when images of handwritten, printed, or typed text are converted into machine-encoded text. Automated recognition of documents, credit cards, car plates and billboards significantly simplifies the way we collect and process data.</p>
<p dir="ltr">The growth of Machine Learning and Convolutional Neural Networks (CNN) has helped text recognition make a huge leap forward. We used CNN in our research to recognize paper receipts from retail stores. The system can be adjusted to process different languages, but we tested it using Russian.</p>
<p dir="ltr">The goal of our project was to develop an app using the client-server architecture for receipt recognition. Let's take a closer look, step by step.</p>
<h2>Preprocessing</h2>
<p dir="ltr">First things first: we rotated the receipt image so that the text lines were horizontally oriented, made the algorithm detect the receipt, and binarized it.</p>
<h3>Rotating Image to Recognize a Receipt</h3>
<p dir="ltr">What we used to recognize a receipt on the image:</p>
<ol>
<li dir="ltr">
<p dir="ltr">Adaptive binarization with a high threshold.</p>
</li>
<li dir="ltr">
<p dir="ltr">Convolutional Neural Network.</p>
</li>
<li dir="ltr">
<p dir="ltr">Haar cascade classifier.</p>
</li>
</ol>
<h4>Using Adaptive Binarization</h4>
<p><img alt="Original receipt view" title="Original receipt view" src="https://dz2cdn1.dzone.com/storage/temp/5613290-img-1.png"></p>
<p dir="ltr">First, we recognized the area on the image that contains the full receipt and almost no background. To achieve this, we rotated the image so that text lines are horizontally oriented.</p>
<p><img alt="Receipt rotation" title="Receipt rotation" src="https://dz2cdn1.dzone.com/storage/temp/5613292-img-2.png"></p>
<p>We used the <code>adaptive_threshold</code>&nbsp;function from the&nbsp;<code>scikit-image</code>&nbsp;library to find the receipt. This function keeps white pixels in areas with a high gradient, while more homogeneous areas turn black. Using this function, we got a homogeneous background with a couple of white pixels. We were searching for them to define the rectangle.</p>
<p><img alt="Identified area containing receipt" title="Identified area containing receipt" src="https://dz2cdn1.dzone.com/storage/temp/5613302-img-3.png"></p>
<h4>Using a CNN</h4>
<p dir="ltr">We decided to find receipt keypoints using a convolutional neural network as we did before for the <a href="https://www.azoft.com/blog/convolutional-neural-networks/" target="_blank">object detection project</a>. We chose the receipt angles as key points. This method performed well, but worse than adaptive binarization with a high threshold.</p>
<p dir="ltr">The CNN was able to define only those angle coordinates relative to the found text. Text to angle orientation varies greatly, meaning this CNN model is not very precise.</p>
<p dir="ltr">See CNN results below.</p>
<p><img alt="Examples of finding the receipt angles using CNN" title="Examples of finding the receipt angles using CNN" src="https://dz2cdn1.dzone.com/storage/temp/5612188-img-4.png"></p>
<h4>Using the Haar Cascade Classifier to Recognize a Receipt</h4>
<p dir="ltr">As a third alternative, we tried the Haar cascade classifier. However, after a week of training the classifier and changing recognition parameters, we didn&rsquo;t get any positive result. Even the CNN performed much better.</p>
<p dir="ltr">Haar cascade classifier results:</p>
<p><img alt="The best results we managed to get from the classifier" title="The best results we managed to get from the classifier" src="https://dz2cdn1.dzone.com/storage/temp/5612197-img-5.png"><img alt="The best results we managed to get from the classifier" title="The best results we managed to get from the classifier" src="https://dz2cdn1.dzone.com/storage/temp/5613211-img-6.png"></p>
<h3>Binarization</h3>
<p>In the end, we used the same <code>adaptive_threshold</code>&nbsp;method for binarization. The window is quite big so that it contains the text as well as the background.</p>
<p><img alt="Receipt binarization" title="Receipt binarization" src="https://dz2cdn1.dzone.com/storage/temp/5613223-img-7.png"></p>
<h2>Text Detection</h2>
<p>Let's cover a few different components of text detection.</p>
<h3>Detecting Text Via Connected Components</h3>
<p dir="ltr">First, we found the connected components using the <code>findContours</code>&nbsp;function from OpenCV. The majority of connected components are characters, but a few are just noisy text fragments left after binarization. We filtered them using maximal/minimal axis.</p>
<p dir="ltr">Then we applied a combining algorithm to compound characters, like <code>:</code>,&nbsp;<code>&#1049;</code>, and <code>=</code>. The characters are then combined into words via a nearest neighbour search. Here is the principle of nearest neighbours method. You need to find the closest neighbor for every character. Then you choose the most appropriate candidate for combination from the right and from the left side. The algorithm is processed until there are no more characters left.</p>
<p><img alt="Finding connected components and forming words (words are highlighted in one color)" title="Finding connected components and forming words (words are highlighted in one color)" src="https://dz2cdn1.dzone.com/storage/temp/5613323-img-8.png"></p>
<p>Then words formed text lines. We used the hypothesis that words in a single line are located at the same height.</p>
<p><img alt="Forming the lines (lines are highlighted in one color)" title="Forming the lines (lines are highlighted in one color)" src="https://dz2cdn1.dzone.com/storage/temp/5613352-img-9.png"></p>
<p>The disadvantage is that this algorithm cannot correctly recognize noisy text.</p>
<h3>Using a Grid for Text Detection</h3>
<p>We found that almost all receipts had monospaced text. So we managed to draw a grid on the receipt and separate characters from each other using grid lines:</p>
<p><img alt="Grid example" title="Grid example" src="https://dz2cdn1.dzone.com/storage/temp/5613369-img-10.png"></p>
<p dir="ltr">The grid simplifies further receipt recognition. A neural network can be applied to every cell of the grid and every character can be easily recognized. The problem of noisy text is gone. Finally, the number of consequent spaces was precisely defined.</p>
<p dir="ltr">We tried the following algorithm to recognize the grid. First, we found connected components in the binary image:</p>
<p><img alt="Finding connected components" title="Finding connected components" src="https://dz2cdn1.dzone.com/storage/temp/5613375-img-11.png"></p>
<p>Then we processed the lower-left angles of the green rectangles and got a set of points specified by coordinates. To determine distortions we used the 2d periodic function:</p>
<p><img alt="Image title" src="https://dz2cdn1.dzone.com/storage/temp/5613381-img-graph.png"></p>
<p><img alt="Graph of the function in formula" title="Graph of the function in formula" src="https://dz2cdn1.dzone.com/storage/temp/5613383-img-12.png"></p>
<p dir="ltr">The main idea behind the receipt grid was to find non-linear geometric distortions with the graph peak points. In other words, we had to find the maximum value sum of this function. Also, we needed to find an optimal distortion.</p>
<p dir="ltr">We parametrized a geometric distortion using the RectBivariateSpline function from the <a href="https://www.scipy.org/" target="_blank">Scipy</a> Python module. We used the minimize Scipy function for optimization.</p>
<p dir="ltr">Here&rsquo;s what we got:</p>
<p><img alt="Incorrectly found grid" title="Incorrectly found grid" src="https://dz2cdn1.dzone.com/storage/temp/5613415-img-14.png"></p>
<p>All in all, this method appeared to be slow and unstable. We decided not to use it again.</p>
<h2>Optical Character Recognition</h2>
<p>Let's deal with recognizing text we found via connected components and recognizing complete words.</p>
<h3>Recognizing Text We Found Via Connected Components</h3>
<p dir="ltr">For text recognition, we used a Convolutional Neural Network (CNN) trained on receipt fonts. As an output, we had probabilities for every character. We took several initial options that together had 99% probability. Then we used a dictionary to check all possible words that can be compiled using these characters. This helped improve the recognition accuracy and eliminate faults caused by similar characters (for example, "&#1047;" and "&#1069;", Cyrillic alphabet).</p>
<p><img alt="Image title" src="https://dz2cdn1.dzone.com/storage/temp/5613447-img-for-31.png"></p>
<p dir="ltr">However, the method&rsquo;s performance is very low when it comes to noisy text recognition.</p>
<h3>Recognizing Complete Words</h3>
<p dir="ltr">It is necessary to recognize complete words when a text is too noisy to recognize it by single characters. We solved this problem using two methods:</p>
<ul>
<li dir="ltr">
<p dir="ltr">LSTM network.</p>
</li>
<li dir="ltr">
<p dir="ltr">Uniform segmentation.</p>
</li>
</ul>
<h4>LSTM Network</h4>
<p dir="ltr">You can read these articles to learn more about <a href="http://arxiv.org/pdf/1506.04395v2.pdf" target="_blank">reading text in deep convolutional sequences</a> and <a href="https://www.researchgate.net/publication/260341307_Can_we_build_language-independent_OCR_using_LSTM_networks" target="_blank">using LSTM networks for language-independent OCR</a>. For this purpose, we used <a href="https://github.com/tmbdev/ocropy" target="_blank">the OCRopus library</a>.</p>
<p dir="ltr">We used monospaced fonts and prepared an artificial sample for training.</p>
<p><img alt="Artificial set" title="Artificial set" src="https://dz2cdn1.dzone.com/storage/temp/5613452-img-15.png"></p>
<p>After the training, we tested our network using a validation set. The test result appeared to be positive. Then we tested it using real receipts. Here is what we got:</p>
<p><img alt="Image title" src="https://dz2cdn1.dzone.com/storage/temp/5613468-img-151.png"></p>
<p dir="ltr">The trained neural network performed well on simple examples. We successfully recognized them before using other methods. The network didn&rsquo;t work for complex cases.</p>
<p dir="ltr">We distorted the training sample and approximated it to the words recognized on receipts.</p>
<p><img alt="Examples of the artificial set" title="Examples of the artificial set" src="https://dz2cdn1.dzone.com/storage/temp/5613475-img-16.png">To avoid network overfitting we stopped the training process several times and continued training the network with the new dataset. Finally, we got the following results:</p>
<p><img alt="Image title" src="https://dz2cdn1.dzone.com/storage/temp/5613511-img-161.png"></p>
<p dir="ltr">Our new network was good at recognizing complex words. But simple word recognition was not so good.</p>
<p dir="ltr">We believe this CNN can perform much better with a single font and minor distortions.</p>
<h4 dir="ltr">Uniform Segmentation</h4>
<p dir="ltr">The receipts font was monospaced, so we decided to split the words by characters. First, we needed to know the width of every character. Thus, the mode of the character width was estimated for every receipt. In the case of a bimodal character width distribution, there are two modes chosen and a specific width is picked for every text line.</p>
<p><img alt="Bimodal distribution of character widths in the receipt" title="Bimodal distribution of character widths in the receipt" src="https://dz2cdn1.dzone.com/storage/temp/5613522-img-17.png"></p>
<p>When we got an approximate character width, we divided the length of the word by the character width to get the approximate number of characters. Then we divided the length of the word by the approximate number of characters, give or take one character:</p>
<p><img alt="Finding the optimal segmentation" title="Finding the optimal segmentation" src="https://dz2cdn1.dzone.com/storage/temp/5613530-img-18.png"></p>
<p>Choosing the best option for division:</p>
<p><img alt="Optimal segmentation" title="Optimal segmentation" src="https://dz2cdn1.dzone.com/storage/temp/5613534-img-19.png"></p>
<p>The accuracy of such segmentation is quite high.</p>
<p><img alt="Algorithm working correctly" title="Algorithm working correctly" src="https://dz2cdn1.dzone.com/storage/temp/5613535-img-20.png"></p>
<p>Sometimes our algorithms performed incorrectly:</p>
<p><img alt="Incorrect performance" title="Incorrect performance" src="https://dz2cdn1.dzone.com/storage/temp/5613538-img-21.png"></p>
<p>Every fragment was processed by a CNN after the segmentation.</p>
<h2>Extracting Meaning From Receipts</h2>
<p>We used regular expressions to find purchases in receipts. There is one feature in common for all the receipts: the price of purchases is written in the XX.XX format, where X is a number. Therefore, it&rsquo;s possible to extract the lines containing purchases. The Individual Taxpayer Number can be found by searching for 10 numbers and tested by the control sum. The Cardholder Name has the format NAME/SURNAME.</p>
<p><img alt="Extracting meaning from receipts" title="Extracting meaning from receipts" src="https://dz2cdn1.dzone.com/storage/temp/5613546-img-22.png"></p>
<h2>Takeaways</h2>
<ol>
<li>
<p dir="ltr">No matter what you choose, an LSTM or another complex method, there is no silver bullet. Some methods are hard to use and not always useful.</p>
</li>
<li>
<p dir="ltr">We'll continue working on the project. For now, the system shows good performance when the recognized text is not noisy.</p>
</li>
</ol>
</div>
</div>
</div>
</div>
</div>
<br>
<br>
<br>
<br>
<br>
<br>
via dzone.com<br>
<br>
August 26, 2022 at 05:39PM
